### ISSUE TITLE
nic_router optimisations

### ISSUE THREAD

**Issue by jschlatow** [genodelabs/genode#4555]
After testing the fixes for #4536 with other scenarios, I noticed that the router slows down if I replace the IP rule with an UDP rule. After a bit of digging and trace recording, I found out that the router performs a `trigger_once()` RPC for every packet. The reason for this it that it updates the timeout of the UDP link. In consequence, the throughput is reduced significantly (a few hundred Mbit/s).

**Comment by jschlatow** [comment]
Commit f2d70a1 applies signal batching for the packet stream interface. I noticed that the router does not register ready_to_{submit,ack} signal handlers. Given that the router discards packets anyway, those handlers are superfluous and indeed reduce the packet throughput. To support this use case, however, commit b0da532 is required to ensure that ack_avail signals are never shadowed by ready_to_submit signals.

**Comment by jschlatow** [comment]
I force pushed a new commit series to my issue4555_nic_router branch:
- ca6121f: applies batching to the packet stream interface (I only changed `get_acked_packets()` to `try_get_acked_packets()` compared to the original commit.)
- f7cdf3b: changes the preference between ack_avail and ready_to_submit signals (unchanged commit)
- 48c66ba: contains the new `Lazy_one_shot_timeout` implementation that resulted from our discussion on the previous commit
- 0c04a3a: avoids frequent calls to `elapsed_us()` on ARM platforms by caching the `curr_time()` result
- a94f569: fixes a minor bug in the packet stream signalling logic

**Comment by jschlatow** [comment]
In additon to the above commit, I found a few other issues for which I do not have fixes at the moment:

1. With elimination of the timer RPC, we are at the point where the checksum calculation becomes the major bottleneck of the NIC router. In particular on the Zynq, this slows down the router by approx. 25%. I think there are two possible improvements: First, we could use incremental checksum calculation as suggested by RFC1071 since the router only updates IP addresses and ports. Second, the checksum is recalculated for every interface in the target domain, yet, I believe all interfaces are supposed to receive the same packet.
2. We should probably eliminate exceptions from the `alloc_packet()` call. In overload situations, i.e. when a sender fills up the entire buffer, we land in situations where the sender receives an ack_avail signal, releases one packet, allocates and sends a packet and fails to allocate a second packet. This is especially relevant if the receiver does not batch ack_avail signals (such as vfs_lwip). In those ping-pong scheduling scenarios, the overhead from catching the `Packet_alloc_failed` exception becomes significant. In case of the NIC router, we will land in an overload situation if the sender is faster than the receiver. The packet buffer will be filled up at some point and the NIC router starts to drop packets. For every dropped packet, we currently have to catch the `Packet_alloc_failed` exception.
3. I'm not sure where the default value of `max_packets_per_signal=32` comes from. I ran some tests with UDP streams and noticed that a larger value leads to a significantly better throughput. The sweet spot for a single stream seems to be 150. For TCP and multiple streams, we could have a totally different picture though.

**Comment by m-stein** [comment]
@jschlatow I fail to understand the rationale behind https://github.com/genodelabs/genode/commit/0c04a3a869943ed5315037652490e5c1ed138098: avoid frequent calls to elapsed_us().

Clearly, ``elapsed_us`` is expensive and shouldn't be called often. However, that's the exact purpose of ``curr_time`` in the timeout framework. There are platforms where ``curr_time`` is expensive, though, namely non-base-hw on ARM. Is this what motivated you? Is it critical for you to run on such platforms? If that's the case, I think we should rather fix the problem timeout framework (see ``repos/base/src/lib/timeout/arm/timer_connection_time.cc``).

**Comment by jschlatow** [comment]
@m-stein What motivated me is that I identified the `curr_time()` call to be quite slow on base-hw. Since this is done for every packet, it becomes a significant bottleneck. I haven't noticed that there is a specific base-hw implementation, though, and therefore assumed it to call `elapsed_us()`. Nevertheless, performing a syscall for every packet iturns out to be still comparably expensive given that we don't need accurate time in the NIC router.

**Comment by m-stein** [comment]
Thanks for the explanation. My feeling is that we should rather discuss how to fix this in the framework. In an offline discussion with Stefan, the idea came up of using the counter reg of the ARM generic timer instead of a syscall. This improvement could be board specific.

We could also think about caching inside the framework given a use-case-specific tolerance.
I sensed already during our talking about lazy timeouts, that we might want to go further and consider a tolerance parameter in the timeout framework itself. If the timeout scheduler considered such a parameter, it could also further improve performance (compared to the lazy timeout approach) by merging different timeouts that fall into the same tolerance range.

**Comment by jschlatow** [comment]
Of course, fixing this in the framework is preferable. I was just looking for a quick and "not-so-dirty" solution.

Regarding your second point, I believe that being able to adjust the tolerance for caching/interpolation will be useful since accuracy requirements can be quite different. However, I'm not sure whether lazy timeout scheduling should be part of the framework given that it appears to be pretty complex and that we don't have a use case apart from the NIC router yet.

**Comment by jschlatow** [comment]
@chelmuth, the Cached_timer commit with the changed message is 07f5626.

**Comment by m-stein** [comment]
@jschlatow Thanks for incorporating my remark regarding the wake-up calls in the NIC router! However, I'm still a bit unsure why we wake up both our sink and all sources. Could you add a clarifying comment before the ``domain.for_each_interface(...);`` block in order to also help others in understanding the code?

**Comment by m-stein** [comment]
@jschlatow Thanks for the fixups! I just added some more suggestions on my 4555_nic_router_optimisations (is in sync with your current issue4555_nic_router branch). Its mostly minor stuff and if you don't like it, I'm fine with reverting it. What do you think?

cced26656c fixup "nic_router: relax timer accuracy for timestamps" (minor style fixes)
a2955eb889 fixup "nic_router: avoid marginal timeout updates" (rename "deadline wanted" "postponed deadline")
3c7487be40 fixup "nic_router: avoid marginal timeout updates" (merge _needs_scheduling())
538e55ca5f fixup "nic_router: avoid marginal timeout updates" (merge _wanted_timeout())
63e80ce8ff fixup "nic_router: avoid marginal timeout updates" (rename "accuracy" "tolerance")
4c677bea13 fixup "nic_router: avoid marginal timeout updates" (minor style fixes)
837d102255 fixup "nic_router: avoid marginal timeout updates" (deadline() const and public)
bb57db0285 fixup "nic_router: avoid marginal timeout updates" (rename timeout.h)

**Comment by chelmuth** [comment]
I'm delighted by the results of this joined effort. Kudos!

**Comment by jschlatow** [comment]
@m-stein Thanks for the fixups. I don't have any objections.

**Comment by m-stein** [comment]
@jschlatow I've implemented your suggestion about updating checksums only once per packet (great remark by the way):

07b7d223dd nic_router: update checksum only once per packet

Is this helpful?

**Comment by m-stein** [comment]
@jschlatow I've implemented your suggestion about updating checksums only once per packet (great remark by the way):

e13ff87090 nic_router: update checksums only once per packet

Is this helpful?

**Comment by jschlatow** [comment]
@m-stein Very nice. I haven't benchmarked the case with multiple interfaces in one domain, yet. It just occured to me when reading the code that there was a scalability issue.

What do you think about the incremental checksum update?

**Comment by m-stein** [comment]
@jschlatow Unfortunately it seems that ca6121f6b5 breaks both ``run/nic_router_dhcp_managed`` and ``run/nic_router_dhcp_unmanaged``. Maybe this is related to the fact that these tests particularly rely on the functioning of link state signals.

**Comment by m-stein** [comment]
@jschlatow I've added a try on incremental checksum updates:
2666b01726 nic_router: incremental L4 checksum updates
e30d1d20cb nic_router: incremental IPv4 checksum updates
1957d48ac4 internet checksum: use signed long for accumulator
4d60019896 internet checksum: no default arg value
a66a1d29b5 nic_router: no IPv4 checksum update on ICMP echo
cd866c4f05 nic_router: no checksum update on IP routing

I hope this is what you had in mind? I would be quite interested in some performance numbers especiialy before, between and after the top-most 2 commits of this series, in case you would be willing to do some measurements :)

**Comment by jschlatow** [comment]
> @jschlatow I've added a try on incremental checksum updates: [2666b01](https://github.com/genodelabs/genode/commit/2666b01726b570dde6c7430c2067a0994878bd61) nic_router: incremental L4 checksum updates [e30d1d2](https://github.com/genodelabs/genode/commit/e30d1d20cb05ec5d5cf3bcc055db4eb30a1ecfed) nic_router: incremental IPv4 checksum updates [1957d48](https://github.com/genodelabs/genode/commit/1957d48ac4bba4952a66ab373aebaca4d87751d3) internet checksum: use signed long for accumulator [4d60019](https://github.com/genodelabs/genode/commit/4d600198969cb6374bfdb2d13ea944434e4e2649) internet checksum: no default arg value [a66a1d2](https://github.com/genodelabs/genode/commit/a66a1d29b579b8d552a068aaae121ec51bb0a8f3) nic_router: no IPv4 checksum update on ICMP echo [cd866c4](https://github.com/genodelabs/genode/commit/cd866c4f05b4e64674296ea59e6fcefb64df1220) nic_router: no checksum update on IP routing
> 
> I hope this is what you had in mind? I would be quite interested in some performance numbers especiialy before, between and after the top-most 2 commits of this series, in case you would be willing to do some measurements :)

That's fantastic. I conducted a few quick measurement on the Zynq since this platform was most impacted by the checksum calculation. Commit 2666b01 increases the throughput of UDP packets (tested with nic_perf) from 510 to 603 Mbit/s (150 packets per signal) resp. from 360 to 400Mbit/s (32 packets per signal). The impact of commit e30d1d2 is neglectable since, unlike UDP, IPv4 only contains a header checksum. When using an IP rule instead of UDP, I get a througput of 640 (150 packets per signal) resp. 420Mbit/s (32 packets per signal).

I'd like to emphasise that my first nic_perf measurement on the Zynq resulted in 117Mbit/s (using an IP rule). After eliminating the exception issue and introducing the Lazy_one_shot_timer, I still got only 280Mbit/s (using an UDP rule). Thus, I consider 400-420Mbit/s quite an achievement.

**Comment by m-stein** [comment]
**DHCP tests**
Wow, you fixed that real quick! So far, the DHCP client should be the only place where the NIC router sends packets by itself.Thanks also for your detailed report regarding your other findings. I can't deal with it currently though. Let's keep it in mind for later.

**Checksums**
Thanks a lot that you reviewed and measured my patches so fast! I'm really happy that the change had such a noticeable impact. The value of 32 packets didn't come from hard evidence. I'd appreciate you raising it to 150 officially and would say that a difference of about 200 Mbit/s is quite an argument for doing so :)

I expected IPv4 checksums to not be that important. However, the router must keep track of  IPv4 header diffs anyway as they influence UDP/TCP/ICMP checksums.

**Comment by chelmuth** [comment]
As there are multiple branches and a couple of commits in the context of this issue that are not merged yet. If any work is ready for staging, I'd appreciate one of you curating a list (or merge_to_staging branch) so I become able to cherry-pick what you approve.

**Comment by m-stein** [comment]
Wonderful :) I've added two very small style fix-ups and an exception handling clean-up that still waited on a topic branch:
66fe8e88f2 fixup "nic_router: avoid marginal timeout updates" (minor style)
827b5a08e1 fixup "nic_router: batch packet stream signals" (minor style)
3f57e45114 nic_router: find ARP cache entries w/o exceptions

**Comment by jschlatow** [comment]
@mstein, thank your for eliminating the packet allocation exceptions. It does not change the overload situations much, though, since the main culprit is the slow receiver. Still it prevents that the router is spending time on exception catching when dropping packets.

**Comment by jschlatow** [comment]
@chelmuth I forgot to adapt the README when changing the default value of max_packets_per_signal. You'll find the fixup in 8cc4c43.

**Comment by alex-ab** [comment]
> @chelmuth I forgot to adapt the README when changing the default value of max_packets_per_signal. You'll find the fixup in [8cc4c43](https://github.com/genodelabs/genode/commit/8cc4c43bd0b5dbb26d1095f2444bf5e0ea9e9e35).

Added to staging.

**Comment by m-stein** [comment]
@alex-ab As I take it from Johannes' last comment, 5edb931c129fdded3fa7284f1ecdee8e13afa78d could be merged as well.

**Comment by m-stein** [comment]
@alex-ab ... if you don't have any objections, of course :)

**Comment by alex-ab** [comment]
> @alex-ab As I take it from Johannes' last comment, [5edb931](https://github.com/genodelabs/genode/commit/5edb931c129fdded3fa7284f1ecdee8e13afa78d) could be merged as well.

done

**Comment by alex-ab** [comment]
On staging several nic_router_* scripts do not work, especially x86. Can someone please check of you @m-stein or @jschlatow.

**Comment by jschlatow** [comment]
@m-stein I started a list of failed tests. I may have missed some tests, so feel free to add them. I'll go top-down through the list and try to fix them. If you have some spare time, I'd suggest you go bottom-up so that we don't work on the same issues.

- [ ] nic_router_flood on x86_32 qemu
- [ ] nic_router_uplinks on x86_64
- [ ] nic_router_uplinks on x86_32
- [ ] nic_router_dhcp_(un)managed on x86_32 with okl4, pistachio and sel4
- [ ] nic_router on x86_* qemu
- [ ] nic_router on x86_64 with base-linux

**Comment by m-stein** [comment]
Thanks for your organizational effort! I'll do as you said.

**Comment by m-stein** [comment]
@jschlatow @alex-ab I just tested nic_router with x86_64/32 lx and qemu (nova/hw) on staging and got all positive results. However, the nic_router test is known to be slightly unstable because of an age-old, and rather complicated issue with lwip (used by the test clients/servers).

**Comment by m-stein** [comment]
@jschlatow @alex-ab The nic_router_dhcp_* problems should be fixed by this:
3fc49bc452 nic_router_dhcp_*: raise timeouts

**Comment by jschlatow** [comment]
@alex-ab @m-stein I've looked into the nic_router_flood test. The recent NIC router modification dug up a problem in the `ipxe_nic_drv`. Even before the changes, we got occasional "slab backend exhausted" errors by the driver but somehow it was able to stabilise. Please find a more detailed explanation and fix in d2f9a59.

**Comment by jschlatow** [comment]
@m-stein Could you have a brief look at the nic_router_uplinks test as well? I don't have a laptop with serial console at hand and am therefore unable to run the test.

**Comment by m-stein** [comment]
@jschlatow @alex-ab Does it have time till Monday? I have to finish some other work till Friday.

**Comment by jschlatow** [comment]
@m-stein sure, take your time.

**Comment by alex-ab** [comment]
> @alex-ab @m-stein I've looked into the nic_router_flood test. The recent NIC router modification dug up a problem in the `ipxe_nic_drv`. Even before the changes, we got occasional "slab backend exhausted" errors by the driver but somehow it was able to stabilise. Please find a more detailed explanation and fix in [d2f9a59](https://github.com/genodelabs/genode/commit/d2f9a59c5a55f166f62876470a5daaaa02d88699).

Great :+1:

**Comment by alex-ab** [comment]
Sure, no time pressure.

**Comment by alex-ab** [comment]
I debugged the last days netperf_*_wifi scripts which are currently not working. After fixing issues due to SMP Linux config enablement of the dde_linux drivers , the wifi driver is back again. The nic_router is now answering pings using the wifi driver as uplink, but the clients of the nic_router are not reachable. I started to revert the nic_router commits up to "nic_router: update checksums only once per packet", then the wifi.run script started to work again (which starts just a simple http server). (see debug branch https://github.com/alex-ab/genode/commits/issue_4555_wifi). @m-stein: as soon as you have time, please check or give some hints what possible is wrong

**Comment by m-stein** [comment]
@jschlatow @alex-ab The broken nic_router_uplinks 32-bit tests are partly broken already on master. Bisecting the nic_router_uplinks 64-bit tests on macho, I found that they break with this commit series:

d9d786a59b fixup "pc: enable SMP and softirq/tasklets in lx_emul" (spinlock defines)
92f3264315 fixup "pc: enable SMP and softirq/tasklets in lx_emul" (irq stack 32-bit)
096998efea vbox6: improve NMI handling
b4422f30c6 Restore dynamic on/off in usb_webcam test
f093b00b06 lx_emul: fix single USB interface claim/release
1905773b68 fixup "pc: enable SMP and softirq/tasklets in lx_emul"
e1c8a3a497 pc: enable SMP and softirq/tasklets in lx_emul
5c2330c1b1 dde_linux: x86 requirements for SMP
0dd33c5395 dde_linux: initialize wait_bit subsystem

However, I cannot get the exact commit as 0dd33c5395 introduces a linker error:
```
ld: wifi.lib.so: undefined reference to `wait_bit_init'
```

That said, it looks as if the problem is not NIC-router-related and I would prefer someone more involved in the DDE-changes to investigate further.

**Comment by alex-ab** [comment]
Have you tried the already prepared https://github.com/alex-ab/genode/commits/issue_4555_wifi branch with run/wifi, as written 6 days ago ? The branch is based on staging and works again after reverting nic_router scripts up to "update checksums only once per packet".

**Comment by m-stein** [comment]
@alex-ab I missed the connection between your issue and the one observed by @jschlatow and just wanted to start with the first reported one. I'll go on with your branch.

**Comment by m-stein** [comment]
@alex-ab Here's a fixup that solves run/wifi on your issue branch and run/nic_router_uplinks on staging for me:

5d66cc2f25 fixup "nic_router: update checksums only once per packet" (correct interface for _router_mac)

PS: For run/wifi I also had to download and add iwlwifi-6000g2b-6.ucode manually (macho). Not sure whether this is relevant.

**Comment by alex-ab** [comment]
> @alex-ab Here's a fixup that solves run/wifi on your issue branch and run/nic_router_uplinks on staging for me:
> 
> [5d66cc2](https://github.com/genodelabs/genode/commit/5d66cc2f25c3066b77bf87995db2877b51aa714b) fixup "nic_router: update checksums only once per packet" (correct interface for _router_mac)
> 
> PS: For run/wifi I also had to download and add iwlwifi-6000g2b-6.ucode manually (macho). Not sure whether this is relevant.

Thanks :+1: , added to staging.

**Comment by nfeske** [comment]
Since the referenced commits entered the master branch today, shall we close this issue?

**Comment by jschlatow** [comment]
@nfeske Yes, I think we can close this issue.
